File: src/commands/clean.ts
import { rm } from "fs/promises";
import { existsSync } from "fs";
import { getReposDir, listSources } from "../lib/git.js";
import { updateAgentsMd, type PackageEntry, type RepoEntry } from "../lib/agents.js";
import type { Registry } from "../types.js";

export interface CleanOptions {
  cwd?: string;
  /** Only clean packages (all registries) */
  packages?: boolean;
  /** Only clean repos */
  repos?: boolean;
  /** Only clean specific registry */
  registry?: Registry;
}

/**
 * Remove all fetched packages and/or repositories
 */
export async function cleanCommand(options: CleanOptions = {}): Promise<void> {
  const cwd = options.cwd || process.cwd();
  const cleanPackages =
    options.packages || (!options.packages && !options.repos);
  const cleanRepos =
    options.repos || (!options.packages && !options.repos && !options.registry);

  let packagesRemoved = 0;
  let reposRemoved = 0;

  // Get current sources
  const sources = await listSources(cwd);

  // Remaining after clean
  let remainingPackages: PackageEntry[] = [...sources.packages];
  let remainingRepos: RepoEntry[] = [...sources.repos];

  // Determine which packages to remove
  let packagesToRemove: PackageEntry[] = [];
  if (cleanPackages) {
    if (options.registry) {
      packagesToRemove = sources.packages.filter(p => p.registry === options.registry);
      remainingPackages = sources.packages.filter(p => p.registry !== options.registry);
    } else {
      packagesToRemove = sources.packages;
      remainingPackages = [];
    }
    packagesRemoved = packagesToRemove.length;
  }

  // Determine which repos to remove
  let reposToRemove: RepoEntry[] = [];
  if (cleanRepos) {
    reposToRemove = sources.repos;
    remainingRepos = [];
    reposRemoved = reposToRemove.length;
  }

  // Extract repo path from full path (removes monorepo subdirectory)
  const extractRepoPath = (fullPath: string): string => {
    const parts = fullPath.split("/");
    if (parts.length >= 4 && parts[0] === "repos") {
      return parts.slice(0, 4).join("/");
    }
    return fullPath;
  };

  // Get unique repo paths from packages being removed
  const packageRepoPaths = new Set(packagesToRemove.map(p => extractRepoPath(p.path)));
  
  // Get repo paths from repos being removed
  const repoRepoPaths = new Set(reposToRemove.map(r => r.path));

  // Get repo paths that are still needed by remaining packages
  const neededRepoPaths = new Set(remainingPackages.map(p => extractRepoPath(p.path)));

  // Combine all repo paths to potentially remove
  const allRepoPaths = new Set([...packageRepoPaths, ...repoRepoPaths]);

  // Remove repos that are no longer needed
  const reposDir = getReposDir(cwd);
  for (const repoPath of allRepoPaths) {
    if (!neededRepoPaths.has(repoPath)) {
      const fullPath = `${cwd}/opensrc/${repoPath}`;
      if (existsSync(fullPath)) {
        await rm(fullPath, { recursive: true, force: true });
      }
    }
  }

  // Clean up empty directories in repos/
  if (existsSync(reposDir)) {
    await cleanupEmptyDirs(reposDir);
  }

  // Summary
  if (cleanPackages) {
    if (options.registry) {
      console.log(`✓ Removed ${packagesRemoved} ${options.registry} package(s)`);
    } else if (packagesRemoved > 0) {
      console.log(`✓ Removed ${packagesRemoved} package(s)`);
    } else {
      console.log("No packages to remove");
    }
  }

  if (cleanRepos) {
    if (reposRemoved > 0) {
      console.log(`✓ Removed ${reposRemoved} repo(s)`);
    } else {
      console.log("No repos to remove");
    }
  }

  const totalRemoved = packagesRemoved + reposRemoved;

  if (totalRemoved > 0) {
    // Update sources.json and AGENTS.md
    await updateAgentsMd({ packages: remainingPackages, repos: remainingRepos }, cwd);

    const totalRemaining = remainingPackages.length + remainingRepos.length;

    if (totalRemaining === 0) {
      console.log("✓ Updated sources.json");
    }
  }

  console.log(`\nCleaned ${totalRemoved} source(s)`);
}

/**
 * Recursively clean up empty directories
 */
async function cleanupEmptyDirs(dir: string): Promise<boolean> {
  const { readdir, rmdir } = await import("fs/promises");
  
  try {
    const entries = await readdir(dir, { withFileTypes: true });
    
    // Recursively clean subdirectories first
    for (const entry of entries) {
      if (entry.isDirectory()) {
        const subdir = `${dir}/${entry.name}`;
        await cleanupEmptyDirs(subdir);
      }
    }
    
    // Check if directory is now empty
    const remaining = await readdir(dir);
    if (remaining.length === 0) {
      await rmdir(dir);
      return true;
    }
  } catch {
    // Ignore errors
  }
  
  return false;
}

File: packages/opensrc/cli/src/commands/clean.rs
use std::fs;

use crate::core::cache::{
    cleanup_empty_dirs, extract_repo_base_path, get_opensrc_dir, get_repos_dir, list_sources,
    write_sources, PackageEntry, RepoEntry,
};
use crate::core::registries::Registry;

pub fn run(
    clean_packages_flag: bool,
    clean_repos_flag: bool,
    registry: Option<Registry>,
) -> Result<(), Box<dyn std::error::Error>> {
    let clean_packages = clean_packages_flag || !clean_repos_flag;
    let clean_repos = clean_repos_flag || (!clean_packages_flag && registry.is_none());

    let (packages, repos) = list_sources();

    let mut remaining_packages: Vec<PackageEntry> = packages.clone();
    let mut remaining_repos: Vec<RepoEntry> = repos.clone();

    let mut packages_to_remove: Vec<PackageEntry> = Vec::new();
    let mut packages_removed = 0usize;

    if clean_packages {
        if let Some(reg) = registry {
            packages_to_remove = packages
                .iter()
                .filter(|p| p.registry == reg)
                .cloned()
                .collect();
            remaining_packages = packages
                .iter()
                .filter(|p| p.registry != reg)
                .cloned()
                .collect();
        } else {
            packages_to_remove = packages.clone();
            remaining_packages = Vec::new();
        }
        packages_removed = packages_to_remove.len();
    }

    let mut repos_to_remove: Vec<RepoEntry> = Vec::new();
    let mut repos_removed = 0usize;

    if clean_repos {
        repos_to_remove = repos.clone();
        remaining_repos = Vec::new();
        repos_removed = repos_to_remove.len();
    }

    // Determine which on-disk paths to remove
    let pkg_paths: std::collections::HashSet<String> = packages_to_remove
        .iter()
        .map(|p| extract_repo_base_path(&p.path))
        .collect();
    let repo_paths: std::collections::HashSet<String> =
        repos_to_remove.iter().map(|r| r.path.clone()).collect();
    let needed_paths: std::collections::HashSet<String> = remaining_packages
        .iter()
        .map(|p| extract_repo_base_path(&p.path))
        .collect();

    let all_paths: std::collections::HashSet<String> =
        pkg_paths.union(&repo_paths).cloned().collect();
    let opensrc_dir = get_opensrc_dir();

    let mut delete_errors = 0usize;
    for repo_path in &all_paths {
        if !needed_paths.contains(repo_path) {
            let full = opensrc_dir.join(repo_path);
            if full.exists() {
                if let Err(e) = fs::remove_dir_all(&full) {
                    eprintln!("Warning: failed to remove {}: {e}", full.display());
                    delete_errors += 1;
                }
            }
        }
    }

    let repos_dir = get_repos_dir();
    if repos_dir.exists() {
        cleanup_empty_dirs(&repos_dir);
    }

    if clean_packages {
        if let Some(reg) = registry {
            println!("✓ Removed {packages_removed} {} package(s)", reg.label());
        } else if packages_removed > 0 {
            println!("✓ Removed {packages_removed} package(s)");
        } else {
            println!("No packages to remove");
        }
    }

    if clean_repos {
        if repos_removed > 0 {
            println!("✓ Removed {repos_removed} repo(s)");
        } else {
            println!("No repos to remove");
        }
    }

    let total = packages_removed + repos_removed;

    if total > 0 {
        write_sources(remaining_packages, remaining_repos)?;
    }

    println!("\nCleaned {total} source(s)");

    if delete_errors > 0 {
        return Err(format!("Failed to remove {delete_errors} path(s) from disk").into());
    }

    Ok(())
}
